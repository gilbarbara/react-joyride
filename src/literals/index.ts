export const ACTIONS = {
  INIT: 'init',
  START: 'start',
  STOP: 'stop',
  RESET: 'reset',
  PREV: 'prev',
  NEXT: 'next',
  GO: 'go',
  CLOSE: 'close',
  SKIP: 'skip',
  UPDATE: 'update',
} as const;

export const EVENTS = {
  TOUR_START: 'tour:start',
  STEP_BEFORE: 'step:before',
  BEACON: 'beacon',
  TOOLTIP: 'tooltip',
  STEP_AFTER: 'step:after',
  TOUR_END: 'tour:end',
  TOUR_STATUS: 'tour:status',
  TARGET_NOT_FOUND: 'error:target_not_found',
  ERROR: 'error',
} as const;

export const LIFECYCLE = {
  INIT: 'init',
  READY: 'ready',
  BEACON: 'beacon',
  TOOLTIP: 'tooltip',
  COMPLETE: 'complete',
  ERROR: 'error',
} as const;

export const ORIGIN = {
  BUTTON_CLOSE: 'button_close',
  BUTTON_PRIMARY: 'button_primary',
  KEYBOARD: 'keyboard',
  OVERLAY: 'overlay',
} as const;

export const STATUS = {
  IDLE: 'idle',
  READY: 'ready',
  WAITING: 'waiting',
  RUNNING: 'running',
  PAUSED: 'paused',
  SKIPPED: 'skipped',
  FINISHED: 'finished',
  ERROR: 'error',
} as const;
