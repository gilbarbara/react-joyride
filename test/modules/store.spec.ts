import { fromPartial } from '@total-typescript/shoehorn';

import createStore from '~/modules/store';

import { LIFECYCLE, STATUS } from '~/literals';

import { standardSteps } from '../__fixtures__/steps';

const mockSyncStore = jest.fn();

describe('store', () => {
  beforeEach(() => {
    mockSyncStore.mockClear();
  });

  describe('without initial values', () => {
    const store = createStore();

    const { close, go, info, next, open, prev, reset, setSteps, start, stop, update } = store;

    it('should have initiated a new store', () => {
      expect(store.constructor.name).toBe('Store');

      expect(info()).toMatchSnapshot();
    });

    it("shouldn't be able to start without steps", () => {
      start();

      expect(info()).toMatchSnapshot();
    });

    it('should ignore all back/forward methods', () => {
      const initialStore = info();

      next();
      expect(info()).toEqual(initialStore);

      prev();
      expect(info()).toEqual(initialStore);

      go(2);
      expect(info()).toEqual(initialStore);
    });

    it('should be able to add steps', () => {
      setSteps(standardSteps);

      expect(info()).toMatchSnapshot();
    });

    it('should handle "prev" but no changes [1st step]', () => {
      prev();

      expect(info()).toMatchSnapshot();
    });

    it(`should handle "update" the lifecycle to ${LIFECYCLE.BEACON}`, () => {
      update({ lifecycle: LIFECYCLE.BEACON });

      expect(info()).toMatchSnapshot();
    });

    it(`should handle "update" the lifecycle to ${LIFECYCLE.TOOLTIP}`, () => {
      update({ lifecycle: LIFECYCLE.TOOLTIP });

      expect(info()).toMatchSnapshot();
    });

    it('should throw an error with `update` with invalid keys', () => {
      expect(() => {
        // @ts-expect-error invalid key
        update({ valid: true, lifecycle: LIFECYCLE.TOOLTIP });
      }).toThrowErrorMatchingSnapshot();
    });

    it('should handle "next" [2nd step]', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "prev" [1st step]', () => {
      prev();

      expect(info()).toMatchSnapshot();
    });

    it('should handle `stop`', () => {
      stop();

      expect(info()).toMatchSnapshot();
    });

    it('should handle `start` [1st step]', () => {
      start();

      expect(info()).toMatchSnapshot();
    });

    it('should handle `stop` again but with `advance`', () => {
      stop(true);

      expect(info()).toMatchSnapshot();
    });

    it('should handle `start` [2nd step]', () => {
      start();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "next" [3rd step]', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "next" [4th step]', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "next" [5th step]', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it(`should handle "update" the lifecycle to ${LIFECYCLE.BEACON}`, () => {
      update({ lifecycle: LIFECYCLE.BEACON });

      expect(info()).toMatchSnapshot();
    });

    it('should handle "next" again but the tour has finished', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "next" again but there\'s no change to the store', () => {
      next();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "reset"', () => {
      reset();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "reset" to restart', () => {
      reset(true);

      expect(info()).toMatchSnapshot();
    });

    it('should handle "start" with custom index and lifecycle', () => {
      start(2);

      expect(info()).toMatchSnapshot();
    });

    it('should handle "go" [2nd step]', () => {
      go(2);

      expect(info()).toMatchSnapshot();
    });

    it('should handle "go" [3rd step]', () => {
      go(1);

      expect(info()).toMatchSnapshot();
    });

    it('should handle "close"', () => {
      close();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "open"', () => {
      open();

      expect(info()).toMatchSnapshot();
    });

    it('should handle "go" with a number higher that the steps length and finish the tour', () => {
      go(10);

      expect(info()).toMatchSnapshot();
    });
  });

  describe('with initial steps', () => {
    const store = createStore({ controlled: false, run: false, steps: standardSteps });

    const { info, update } = store;

    it('should have initiated a new store', () => {
      expect(store.constructor.name).toBe('Store');

      expect(info()).toMatchSnapshot();
    });

    it('should handle listeners', () => {
      store.addListener(mockSyncStore);

      update({ status: STATUS.READY });
      update({ status: STATUS.READY });
      expect(mockSyncStore).toHaveBeenCalledTimes(1);

      update({ status: STATUS.IDLE });
      update({ status: STATUS.READY });

      expect(mockSyncStore).toHaveBeenCalledTimes(3);
    });
  });

  describe('with popper', () => {
    const store = createStore();
    const popperData = { placement: 'top' } as const;

    it('should set/get both poppers', () => {
      store.setPopper('beacon', fromPartial(popperData));
      expect(store.getPopper('beacon')).toEqual(popperData);

      store.setPopper('tooltip', fromPartial(popperData));
      expect(store.getPopper('tooltip')).toEqual(popperData);
    });

    it('should clear both poppers', () => {
      expect(store.getPopper('beacon')).toEqual(popperData);
      expect(store.getPopper('tooltip')).toEqual(popperData);

      store.cleanupPoppers();
      expect(store.getPopper('beacon')).toBeNull();
      expect(store.getPopper('tooltip')).toBeNull();
    });
  });
});
