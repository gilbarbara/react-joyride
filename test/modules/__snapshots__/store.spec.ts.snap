// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`store with initial steps should have initiated a new store 1`] = `
{
  "action": "init",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "ready",
}
`;

exports[`store without initial values should be able to add steps 1`] = `
{
  "action": "start",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "close" 1`] = `
{
  "action": "close",
  "controlled": false,
  "index": 2,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "go" [2nd step] 1`] = `
{
  "action": "go",
  "controlled": false,
  "index": 2,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "go" [3rd step] 1`] = `
{
  "action": "go",
  "controlled": false,
  "index": 1,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "go" with a number higher that the steps length and finish the tour 1`] = `
{
  "action": "go",
  "controlled": false,
  "index": 6,
  "lifecycle": "init",
  "size": 6,
  "status": "finished",
}
`;

exports[`store without initial values should handle "next" [2nd step] 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 1,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "next" [3rd step] 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 2,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "next" [4th step] 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 3,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "next" [5th step] 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 4,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "next" again but the tour has finished 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 5,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "next" again but there's no change to the store 1`] = `
{
  "action": "next",
  "controlled": false,
  "index": 6,
  "lifecycle": "init",
  "size": 6,
  "status": "finished",
}
`;

exports[`store without initial values should handle "open" 1`] = `
{
  "action": "update",
  "controlled": false,
  "index": 2,
  "lifecycle": "tooltip",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "prev" [1st step] 1`] = `
{
  "action": "prev",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "prev" but no changes [1st step] 1`] = `
{
  "action": "prev",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "reset" 1`] = `
{
  "action": "reset",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "ready",
}
`;

exports[`store without initial values should handle "reset" to restart 1`] = `
{
  "action": "reset",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "start" with custom index and lifecycle 1`] = `
{
  "action": "start",
  "controlled": false,
  "index": 2,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "update" the lifecycle to beacon 1`] = `
{
  "action": "update",
  "controlled": false,
  "index": 0,
  "lifecycle": "beacon",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "update" the lifecycle to beacon 2`] = `
{
  "action": "update",
  "controlled": false,
  "index": 4,
  "lifecycle": "beacon",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle "update" the lifecycle to tooltip 1`] = `
{
  "action": "update",
  "controlled": false,
  "index": 0,
  "lifecycle": "tooltip",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle \`start\` [1st step] 1`] = `
{
  "action": "start",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle \`start\` [2nd step] 1`] = `
{
  "action": "start",
  "controlled": false,
  "index": 1,
  "lifecycle": "init",
  "size": 6,
  "status": "running",
}
`;

exports[`store without initial values should handle \`stop\` 1`] = `
{
  "action": "stop",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 6,
  "status": "paused",
}
`;

exports[`store without initial values should handle \`stop\` again but with \`advance\` 1`] = `
{
  "action": "stop",
  "controlled": false,
  "index": 1,
  "lifecycle": "init",
  "size": 6,
  "status": "paused",
}
`;

exports[`store without initial values should have initiated a new store 1`] = `
{
  "action": "init",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 0,
  "status": "idle",
}
`;

exports[`store without initial values should throw an error with \`update\` with invalid keys 1`] = `"State is not valid. Valid keys: action, index, lifecycle, status"`;

exports[`store without initial values shouldn't be able to start without steps 1`] = `
{
  "action": "start",
  "controlled": false,
  "index": 0,
  "lifecycle": "init",
  "size": 0,
  "status": "waiting",
}
`;
